iml-wasm-components/iml-input/src/lib.rs

```diff
@@ -0,0 +1,118 @@
+extern crate combine;
+extern crate seed;
+
+use combine::stream::easy::{Error, Errors, Info};
+use seed::{div, input, prelude::*};
+
+use iml_tooltip::TooltipPlacement;
+
+pub struct Model<T> {
+    pub value: Result<T, String>,
+    pub tooltip: iml_tooltip::Model,
+}
+
+impl<T> Default for Model<T> {
+    fn default() -> Self {
+        Self {
+            value: Err(String::new()),
+            tooltip: iml_tooltip::Model {
+                error_tooltip: true,
+                open: true,
+                placement: TooltipPlacement::Bottom,
+                ..Default::default()
+            },
+        }
+    }
+}
+
+#[derive(Clone)]
+pub enum Msg {
+    FlipTooltip,
+    InputChanged(String),
+}
+
+pub type ParserFn<T> = fn(_: &str) -> Result<T, Errors<char, &str, usize>>;
+
+fn format_err(err: &Errors<char, &str, usize>) -> String {
+    let mut un = vec![];
+    let mut ex = vec![];
+
+    fn format_info(i: &Info<char, &str>) -> String {
+        match i {
+            Info::Token(t) => {
+                if *t != '"' {
+                    format!("\"{}\"", t)
+                } else {
+                    format!("{}", t)
+                }
+            }
+            Info::Static(s) => String::from(*s),
+            Info::Owned(s) => String::from(s),
+            Info::Range(r) => String::from(*r),
+        }
+    }
+
+    for x in &err.errors {
+        match x {
+            Error::Unexpected(i) => {
+                un.push(format_info(i));
+            }
+            Error::Expected(i) => {
+                ex.push(format_info(i));
+            }
+            _ => {}
+        }
+    }
+
+    let mut s = format!("Unexpected {}", un.join(", "));
+    if ex.len() > 0 {
+        s.push_str(&format!(". Expecting {}", ex.join(", ")));
+    }
+    s
+}
+
+pub fn update<T>(msg: Msg, m: &mut Model<T>, parse: ParserFn<T>) {
+    match msg {
+        Msg::FlipTooltip => {
+            m.tooltip.placement = match m.tooltip.placement {
+                TooltipPlacement::Bottom => TooltipPlacement::Top,
+                _ => TooltipPlacement::Bottom,
+            }
+        }
+        Msg::InputChanged(txt) => match parse(&txt) {
+            Ok(r) => {
+                m.value = Ok(r);
+            }
+            Err(e) => {
+                m.value = Err(format_err(&e));
+            }
+        },
+    }
+}
+
+pub fn view<T>(m: &Model<T>) -> Node<Msg> {
+    let mut input = input![];
+    input.add_listener(input_ev(Ev::Input, Msg::InputChanged));
+    input.add_class("form-control");
```

Comment:
nit, we should prefer the macro style for building the view when we can use it.

This can be:

```rust
let input = input![
        class!["form-control"],
        input_ev(Ev::Input, Msg::InputChanged)
    ];
```